algorithms/branching.py

```diff
@@ -0,0 +1,93 @@
+        if not count_flag:
+            return False
+        return 0
+    if is_bijected:
+        # If the graph is balanced and bijected, the graph has exactly one isomorphism
+        if not count_flag:
+            return True
+        return 1
+
+    G_copy = G.copy()
+    colors, max_colornum = get_colors(G_copy)
+
+    # Choose the color C with at least 4 vertices of that color in the graph
+    for key in colors:
+        if len(colors[key]) >= 4:
+            C = key
+            continue
+
+    # Choose the first occurring vertex with color C in the list of vertices of the first graph
+    g0 = []
+    for v in colors[C]:
+        if v.graph_label == 1:
+            g0.append(v)
+    x = g0[0]
+    # Change the color of this vertex to a new color and append it to the list of fixed vertices for the first graph
+    x.colornum = max_colornum + 1
+    D.append(x)
+    # Create branches for all the possible fixed pairs of vertices for the chosen color
+    return __branching(G_copy, colors, C, D.copy(), I.copy(), count_flag)
+
+
+def __branching(G: 'Graph', colors: 'Dict[Int, List[Vertex]]', C: 'Int', D: 'List[Vertex]', I: 'List[Vertex]', count_flag: 'Bool'):
+    """
+    Creates branches of the graph (disjoint union of two graphs) and count the amount of isomorphisms for those graphs.
+    In one graph, one vertex of the color group is fixed. For each of the vertices in the other graph, a branch is
+    created fixing that vertex.
+    :param G: The graph (disjoint union of two graphs) to branch
+    :param colors: The colors present in the graph (disjoint union of two graphs) and its corresponding list of vertices
+    :param C: The chosen color group to create branches for
+    :param D: The list of vertices in one of the graphs that is fixed
+    :param I: The list of vertices in the other graph that is fixed
+    :param count_flag: Whether or not the amount of isomorphisms should be return or whether or not the graph has an
+    isomorphism
+    :return: The number of isomorphisms if 'count_flag' is True or whether or not the graph has at least one
+    isomorphism if 'count_flag' is False
+    """
+    # Create the list of vertices in the other graph with color C
+    g1 = []
+    for v in colors[C]:
+        if v.graph_label == 2:
+            g1.append(v)
+
+    num_isomorphisms = 0
+
+    # For each of the vertices in the list of vertices with color C, fix the vertex and change its color to the new
+    # color and determine the amount of isomorphisms for the resulting graph
+    for y0 in g1:
+        G_copy = G.copy()
+        D_copy = D.copy()
+        I_copy = I.copy()
+        colors_copy, max_colornum = get_colors(G_copy)
+        for y in colors_copy[C]:
+            if y.graph_label == 2 and y.label == y0.label:
+                y.colornum = max_colornum
+                I_copy.append(y)
+                num_isomorphisms += count_isomorphisms(G_copy, D_copy, I_copy, count_flag)
```

Comment:
Wat gebeurt er als we diep in een recursieve berekening zitten en methode 'branching' returned True, door regel 91 en daardoor returned count_isomorphisms op regel 90 bijvoorbeeld True. Gaat dat dan goed? Is num_isomorphisms op regel 91 dan groter dan 0 ??

Reply:
Als er ergens in de meerdere malen dat branching wordt aangeroepen op regel 91 True wordt gereturnd, dan stopt die ronde van count_isomorphisms met true. Dan moeten inderdaad ook al die andere calls nog wel stoppen. Dit gaat volgens mij toevallig goed, omdat als je in de Python console probeert `x = 1` en dan `x += True`, dan komt er  `x = 2` uit. En dan wordt num_isomorphisms dus altijd weer groter dan nul en zo blijft er True doorgegeven worden. Volgens mij hebben we gewoon geluk dat dat zo werkt 😅

Reply:
Als ik in de console interpreter het volgende stop:

> True > 0

Is het resultaat

> True

Reply:
Ik ben inderdaad bang dat het gewoon toevallig goed werkt, maar als True > 0 inderdaad True teruggeeft en False > 0 geeft False. Gaan we dit dan gebruiken of gaan we onze methode aanpassen op iets wat iets meer logisch is? Nu ik dit bovenste weet denk ik, nou misschien moeten we dat dan ook maar gebruiken dan.

Reply:
Ik vind het lastigst dat deze functie return een 'Bool' of 'Int' kan zijn.
Maar hij werkt wel.

Reply:
> 
> 
> Wat gebeurt er als we diep in een recursieve berekening zitten en methode 'branching' returned True, door regel 91 en daardoor returned count_isomorphisms op regel 90 bijvoorbeeld True. Gaat dat dan goed? Is num_isomorphisms op regel 91 dan groter dan 0 ??

er staat in de if statement:
if not count_flag and num_isomorphisms > 0:
    return True

De grootte van num_isomorphisms doet er in dit geval niet toe, aangezien je alleen returned als je niet in num_isomorphisms geinteresseerd bent.
Dit codeblok r.30 en r.31 is het enige return True statement dat er nog is.
Volgensmij is het overbodig.

        if not count_flag:
            return True

r.32 returnt namelijk num_isomorphisms = 1
        return 1

Zou je het verwijderen van regel 30 en 31 kunnen testen ?

Reply:
> 
> Booleans (bool)
> 
>     These represent the truth values False and True. The two objects representing the values False and True are the only Boolean objects. The Boolean type is a subtype of the integer type, and Boolean values behave like the values 0 and 1, respectively, in almost all contexts, the exception being that when converted to a string, the strings "False" or "True" are returned, respectively.

https://docs.python.org/release/3.0.1/reference/datamodel.html#the-standard-type-hierarchy

Reply:
Ik denk dat we het voor nu prima zo kunnen gebruiken.